File: packages/sdk/src/processes.ts
/**
 * Managed process namespace for starting and stopping long-running MOD command processes.
 *
 * Provides lifecycle management for long-running processes with exit detection
 * via the signals WebSocket. Each {@link ProcessHandle} is `AsyncDisposable` for
 * automatic cleanup.
 *
 * @example
 * ```typescript
 * import { processes } from "@hmcs/sdk";
 *
 * const proc = await processes.start({
 *   command: "@hmcs/persona:default-behavior",
 *   args: ["persona-1"],
 * });
 *
 * proc.onExit((info) => {
 *   console.log(`Process ${info.reason}: exit code ${info.exitCode}`);
 * });
 *
 * // Stop and clean up
 * await proc.stop();
 * ```
 *
 * @packageDocumentation
 */

import { host } from './host';
import { type Subscription, signals } from './signals';

/** Information about why a managed process exited unexpectedly. */
export interface ProcessExitInfo {
  /** Process exit code, or null if killed by a signal. */
  exitCode: number | null;
  /** Unix signal name if killed by a signal. */
  signal: string | null;
  /** Exit reason: `"exited"` for clean exit (code 0), `"crashed"` for non-zero/signal. */
  reason: 'exited' | 'crashed';
}

/** Information about a running managed process. */
export interface ProcessInfo {
  /** Unique handle identifier. */
  handleId: string;
  /** The MOD command reference that was started. */
  command: string;
  /** Arguments passed to the process. */
  args: string[];
  /** OS process ID. */
  pid: number;
  /** ISO 8601 timestamp of when the process was started. */
  startedAt: string;
}

/** Signal payload shape for the process:exited channel. */
interface ExitSignalPayload {
  handleId: string;
  command: string;
  exitCode: number | null;
  signal: string | null;
  reason: 'exited' | 'crashed';
}

/**
 * Handle to a running managed process.
 *
 * Implements `AsyncDisposable` for automatic cleanup via `await using`.
 * Disposing stops the process (if still running) and disconnects the exit listener.
 *
 * @example
 * ```typescript
 * {
 *   await using proc = await processes.start({ command: "my-mod:script" });
 *   proc.onExit((info) => console.log("Crashed!", info));
 *   // proc is automatically stopped and cleaned up at end of block
 * }
 * ```
 */
export class ProcessHandle implements AsyncDisposable {
  /** Unique handle identifier for this process. */
  readonly handleId: string;

  private disposed = false;
  private exited = false;
  private exitCallbacks: Array<(info: ProcessExitInfo) => void> = [];
  private subscription: Subscription;

  constructor(handleId: string) {
    this.handleId = handleId;
    this.subscription = signals.stream<ExitSignalPayload>('process:exited', (payload) => {
      if (payload.handleId !== this.handleId) return;
      if (this.disposed) return;

      this.exited = true;
      const info: ProcessExitInfo = {
        exitCode: payload.exitCode,
        signal: payload.signal,
        reason: payload.reason,
      };
      for (const cb of this.exitCallbacks) {
        try {
          cb(info);
        } catch (e) {
          console.error('Error in onExit callback:', e);
        }
      }
      this.subscription.close();
    });
  }

  /**
   * Register a callback for unexpected process termination.
   *
   * Multiple callbacks can be registered (addEventListener-style).
   * Callbacks are NOT invoked when the process is stopped via {@link stop}
   * or `[Symbol.asyncDispose]`.
   *
   * @param callback - Function called with exit information
   *
   * @example
   * ```typescript
   * proc.onExit((info) => {
   *   if (info.reason === "crashed") {
   *     console.error(`Crash! Exit code: ${info.exitCode}`);
   *   }
   * });
   * ```
   */
  onExit(callback: (info: ProcessExitInfo) => void): void {
    this.exitCallbacks.push(callback);
  }

  /**
   * Stop the process and clean up the exit listener.
   *
   * Idempotent — safe to call multiple times or on an already-exited process.
   *
   * If the process is still running, this sends a stop request to the engine
   * and awaits shutdown. If the process has already exited, this only
   * disconnects the exit listener. `onExit` callbacks are NOT invoked.
   *
   * @example
   * ```typescript
   * const proc = await processes.start({ command: "my-mod:script" });
   * // ...later
   * await proc.stop();
   * ```
   */
  async stop(): Promise<void> {
    if (this.disposed) return;
    this.disposed = true;
    this.subscription.close();

    if (!this.exited) {
      try {
        await host.deleteMethod(host.createUrl(`processes/${this.handleId}`));
      } catch {
        // 404 = already exited — idempotent success
      }
    }
  }

  /** Alias for {@link stop}. Enables `await using` syntax. */
  [Symbol.asyncDispose] = (): Promise<void> => this.stop();
}

export namespace processes {
  /**
   * Start a long-running MOD command process.
   *
   * Returns a {@link ProcessHandle} for lifecycle management.
   * The handle is `AsyncDisposable` — use `await using` for automatic cleanup.
   *
   * @param params - Process start parameters
   * @returns A handle to the running process
   *
   * @example
   * ```typescript
   * const proc = await processes.start({
   *   command: "@hmcs/persona:default-behavior",
   *   args: ["persona-1"],
   * });
   * await proc.stop();
   * ```
   */
  export async function start(params: {
    /** Full MOD command reference (`mod-name:bin-name`). */
    command: string;
    /** Arguments forwarded to the process. */
    args?: string[];
  }): Promise<ProcessHandle> {
    const response = await host.post(host.createUrl('processes/start'), {
      command: params.command,
      args: params.args ?? [],
    });
    const { handleId } = (await response.json()) as { handleId: string };
    return new ProcessHandle(handleId);
  }

  /**
   * List all running managed processes.
   *
   * @returns Array of process information objects
   *
   * @example
   * ```typescript
   * const running = await processes.list();
   * for (const p of running) {
   *   console.log(`${p.command} (pid ${p.pid}) since ${p.startedAt}`);
   * }
   * ```
   */
  export async function list(): Promise<ProcessInfo[]> {
    const response = await host.get(host.createUrl('processes'));
    return (await response.json()) as ProcessInfo[];
  }
}

File: engine/crates/homunculus_api/src/processes.rs
//! Managed processes API.
//!
//! Provides async wrappers for starting, stopping, and listing long-running
//! MOD command processes. Includes an exit detection system that emits
//! `process:exited` signals when processes terminate unexpectedly.

use crate::api;
use crate::error::{ApiError, ApiResult};
use crate::signals::SignalsChannels;
use bevy::prelude::*;
use bevy_flurx::prelude::*;
use homunculus_core::prelude::{HomunculusConfig, ModRegistry};
use homunculus_mod::managed_process::{MAX_PROCESSES, ManagedProcess};
use homunculus_utils::runtime::RuntimeResolver;
use homunculus_mod::node_process::NodeProcessHandle;
use serde::{Deserialize, Serialize};
use std::time::Duration;

api!(
    /// Provides access to the managed processes API.
    ProcessesApi
);

/// Request to start a managed process.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct StartProcessRequest {
    /// Full MOD command reference (`mod-name:bin-name`).
    pub command: String,
    /// Arguments forwarded to the process as CLI args.
    #[serde(default)]
    pub args: Vec<String>,
}

/// Response from starting a managed process.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct StartProcessResponse {
    /// Unique handle ID for the started process.
    pub handle_id: String,
}

/// Information about a running managed process.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct ProcessInfo {
    /// Unique handle identifier.
    pub handle_id: String,
    /// Full MOD command reference.
    pub command: String,
    /// Arguments passed to the process.
    pub args: Vec<String>,
    /// OS process ID.
    pub pid: u32,
    /// ISO 8601 timestamp of when the process was started.
    pub started_at: String,
}

impl ProcessesApi {
    /// Start a new managed process.
    pub async fn start(&self, req: StartProcessRequest) -> ApiResult<StartProcessResponse> {
        self.0
            .schedule(move |task| async move {
                task.will(Update, once::run(start_process).with(req)).await
            })
            .await?
    }

    /// Stop a managed process by handle ID.
    ///
    /// The shutdown is fully non-blocking for the Bevy main thread:
    /// entity lookup and despawn happen in a one-shot system, then the
    /// actual SIGTERM → grace → SIGKILL sequence runs in a tokio blocking task.
    pub async fn stop(&self, handle_id: String) -> ApiResult<()> {
        self.0
            .schedule(move |task| async move {
                // Step 1: Take NodeProcessHandle ownership and despawn entity.
                let mut handle = task
                    .will(Update, once::run(take_and_despawn).with(handle_id))
                    .await?;

                // Step 2: Async shutdown in tokio (does NOT block Bevy).
                task.will(
                    Update,
                    side_effect::tokio::spawn(async move {
                        tokio::task::spawn_blocking(move || {
                            handle.shutdown(Duration::from_secs(2));
                        })
                        .await
                        .ok();
                    }),
                )
                .await;

                Ok(())
            })
            .await?
    }

    /// List all running managed processes.
    pub async fn list(&self) -> ApiResult<Vec<ProcessInfo>> {
        self.0
            .schedule(move |task| async move { task.will(Update, once::run(list_processes)).await })
            .await
    }
}

fn start_process(
    In(req): In<StartProcessRequest>,
    mut commands: Commands,
    registry: Res<ModRegistry>,
    config: Res<HomunculusConfig>,
    runtime: Res<RuntimeResolver>,
    existing: Query<&ManagedProcess>,
) -> ApiResult<StartProcessResponse> {
    if existing.iter().count() >= MAX_PROCESSES {
        return Err(ApiError::TooManyRequests(format!(
            "Maximum {MAX_PROCESSES} concurrent managed processes"
        )));
    }

    let result = homunculus_mod::managed_process::spawn_managed_process(
        &mut commands,
        &registry,
        &config,
        &runtime,
        &req.command,
        req.args,
    )
    .map_err(ApiError::InvalidInput)?;

    Ok(StartProcessResponse {
        handle_id: result.handle_id,
    })
}

/// Takes [`NodeProcessHandle`] ownership from the entity and despawns it.
///
/// This is an exclusive system (`&mut World`) because `entity_mut().take::<T>()`
/// requires immediate component removal (not deferred via `Commands`).
fn take_and_despawn(In(handle_id): In<String>, world: &mut World) -> ApiResult<NodeProcessHandle> {
    let entity = world
        .query::<(Entity, &ManagedProcess)>()
        .iter(world)
        .find(|(_, m)| m.handle_id == handle_id)
        .map(|(e, _)| e)
        .ok_or(ApiError::EntityNotFound)?;
    let handle = world
        .entity_mut(entity)
        .take::<NodeProcessHandle>()
        .ok_or(ApiError::EntityNotFound)?;
    world.despawn(entity);
    Ok(handle)
}

fn list_processes(query: Query<&ManagedProcess>) -> Vec<ProcessInfo> {
    query
        .iter()
        .map(|m| ProcessInfo {
            handle_id: m.handle_id.clone(),
            command: m.command.clone(),
            args: m.args.clone(),
            pid: m.pid,
            started_at: m.started_at.to_rfc3339(),
        })
        .collect()
}

/// Plugin that registers the exit detection system.
pub(crate) struct ProcessesApiPlugin;

impl Plugin for ProcessesApiPlugin {
    fn build(&self, app: &mut App) {
        app.add_systems(Update, check_process_exits);
    }
}

/// Polls `try_wait_exited()` on all managed processes each frame.
/// Emits `process:exited` signal and despawns the entity when a process has exited.
fn check_process_exits(
    mut commands: Commands,
    mut query: Query<(Entity, &ManagedProcess, &mut NodeProcessHandle)>,
    mut channels: ResMut<SignalsChannels>,
) {
    for (entity, managed, mut handle) in query.iter_mut() {
        if let Some(status) = handle.try_wait_exited() {
            let exit_code = status.code();
            #[cfg(unix)]
            let signal = {
                use std::os::unix::process::ExitStatusExt;
                status.signal().map(|s| format!("{s}"))
            };
            #[cfg(not(unix))]
            let signal: Option<String> = None;

            let reason = if exit_code == Some(0) {
                "exited"
            } else {
                "crashed"
            };

            let payload = serde_json::json!({
                "handleId": managed.handle_id,
                "command": managed.command,
                "exitCode": exit_code,
                "signal": signal,
                "reason": reason,
            });

            if let Err(e) = channels.send_blocking("process:exited", payload) {
                error!(
                    "Failed to send process:exited signal for {}: {e}",
                    managed.handle_id
                );
            }

            info!(
                "Managed process '{}' {reason} (code: {exit_code:?})",
                managed.handle_id
            );
            commands.entity(entity).despawn();
        }
    }
}
